lib/PromAlertProxy/Target/Email.pm

```diff
@@ -0,0 +1,102 @@
+package PromAlertProxy::Target::Email;
+
+use 5.028;
+use Moo;
+use experimental qw(signatures);
+
+with 'PromAlertProxy::Target';
+
+use Types::Standard qw(Object Str HashRef);
+use Type::Utils qw(role_type class_type);
+use Type::Params qw(compile);
+
+use Module::Runtime qw(require_module);
+
+use Email::Stuffer;
+use Template::Tiny;
+
+use PromAlertProxy::Logger '$Logger';
+
+has from => (
+  is       => 'ro',
+  isa      => Str, # XXX EmailAddress
+  required => 1,
+);
+
+has to => (
+  is       => 'ro',
+  isa      => Str, # XXX EmailAddress
+  required => 1,
+);
+
+has transport => (
+  is       => 'ro',
+  isa      => Str,
+  required => 1,
+);
+
+has transport_args => (
+  is      => 'ro',
+  isa     => HashRef,
+  default => sub { {} },
+);
+
+has _transport => (
+  is      => 'lazy',
+  isa     => role_type('Email::Sender::Transport'),
+  default => sub ($self) {
+    my $transport_class = 'Email::Sender::Transport::'.$self->transport;
+    require_module($transport_class);
+    $transport_class->new($self->transport_args);
+  },
+);
+
+sub raise ($self, $alert) {
+  state $check = compile(Object, class_type('PromAlertProxy::PromAlert'));
+  $check->(@_);
+
+  Email::Stuffer->from($self->from)
+                ->to($self->to)
+                ->subject($alert->name.': '.$alert->summary)
+                ->text_body($self->_body_for($alert))
+                ->header('In-Reply-To' => '<'.$alert->key.'@promalertproxy>')
+                ->transport($self->_transport)
+                ->send_or_die;
+
+  return;
+}
+
+sub _body_for ($self, $alert) {
+  state $template = do { local $/; <DATA> };
+
+  my $output;
+  Template::Tiny->new->process(\$template, {
+    key              => $alert->key,
+    name             => $alert->name,
+    summary          => $alert->summary,
+    description      => $alert->description,
+    labels           => $alert->labels,
+    annotations      => $alert->annotations,
+    dump_labels      =>
+      do { my $d = $alert->labels; join("\n", map { "$_: $d->{$_}" } sort keys %$d) },
+    dump_annotations =>
+      do { my $d = $alert->annotations; join("\n", map { "$_: $d->{$_}" } sort keys %$d) },
```

Comment:
I do not know of pythonic-d!

(I like `do`, totally hate this whole thing, and am leaving it).

Reply:
I just meant calling it $d (presumably for "dictionary") rather than $h or whatever.

BUT SINCE I'M HERE ANYWAY. What I would do is this:

```perl 
sub format_hash ($what) {  join("\n", map { "$_: $hash->{$_}" } sort keys %$hash) } }

Template::Tiny->new->defile($template, {
  ...
  dump_annotations => format_hash($self->annotations),
  dump_labels => format_hash($self->labels)
})
```